cabal-install/src/Distribution/Client/Types/SourcePackageDb.hs

```diff
@@ -21,3 +26,38 @@ data SourcePackageDb = SourcePackageDb
   deriving (Eq, Generic)
 
 instance Binary SourcePackageDb
+
+-- | Does a case-sensitive search by package name and a range of versions.
+--
+-- We get back any number of versions of the specified package name, all
+-- satisfying the version range constraint.
+--
+-- Additionally, `preferred-versions` (such as version deprecation) are
+-- honoured in this lookup, which is the only difference to
+-- 'PackageIndex.lookupDependency'
+lookupDependency :: SourcePackageDb -> PackageName -> VersionRange -> [UnresolvedSourcePackage]
+lookupDependency sourceDb pname version =
+  filterPreferredVersions pref $ PackageIndex.lookupDependency (packageIndex sourceDb) pname version
+  where
+    pref = Map.lookup pname (packagePreferences sourceDb)
+
+
+-- | Does a case-sensitive search by package name.
+--
+-- Additionally, `preferred-versions` (such as version deprecation) are
+-- honoured in this lookup, which is the only difference to
+-- 'PackageIndex.lookupPackageName'
+lookupPackageName :: SourcePackageDb -> PackageName -> [UnresolvedSourcePackage]
```

Comment:
Also, afaict there should be other codepaths that would benefit from these functions (e.g. list). Should I review those in a follow-up PR to unify logic?

Reply:
IMHO, yes, please.